### ISSUE TITLE
CricleCI Plugin - Feature: See all jobs + include approving deployments

### ISSUE THREAD

**Issue by igor-radulovic** [backstage/backstage#9428]
## Feature Suggestion

The CircleCI plugin is a great tool to visualize recent builds and provides the option to re-run builds. What I and my team feel is missing is visualization of jobs that haven't started yet and the option to handle manual approvement steps. To better explain this feature/improvment: In our Circle config, we have certain jobs that are dependent on other jobs, lets say a deployment is dependent on a build job, those are stacked together into one workflow in CircleCI while they are split in the plugin. This makes it more difficult to see what jobs are dependant on each other, and also see upcoming jobs in the workflow. This also applies to the approval deployment. Lets say we first deploy to a play environment and to continue deploying it up the the next environment level e.g. development, a manual approvment has to be done for the workflow to continue (as can be seen in one of the pictures below)

## Possible Implementation

[https://circleci.com/blog/introducing-circleci-api-v2/](https://circleci.com/blog/introducing-circleci-api-v2/)
From what I can see in API V2, it is possible to retrieve jobs from a workflow.

## Context

![image](https://user-images.githubusercontent.com/92377369/153155414-70505d7f-cfb0-4b72-9d3f-6592060060dd.png)

![image](https://user-images.githubusercontent.com/92377369/153062508-5348167b-ff46-444d-88db-ba191987a699.png)

**Comment by Rugvip** [comment]
Sounds like a great idea! :grin: Is this something that you plan to work on @igor-radulovic?

**Comment by igor-radulovic** [comment]
@Rugvip I'll have to leave this on hold for now, unfortunately. We have some decisions to make, and if we end up on Circle I can take a look at it :) 

Also since this requires using API V2, the task becomes a bit larger [https://github.com/backstage/backstage/issues/9361](https://github.com/backstage/backstage/issues/9361)

**Comment by jithenshriyan** [comment]
> @Rugvip I'll have to leave this on hold for now, unfortunately. We have some decisions to make, and if we end up on Circle I can take a look at it :)
> 
> Also since this requires using API V2, the task becomes a bit larger #9361

I released an SDK for the CircleCI V2 API this week covering workflows, jobs, pipelines, and context endpoints. We could use my SDK to migrate the service to the V2 API. Everything has Typescript declarations as well so we can plug it in.

Here's the repo: https://github.com/jithenshriyan/circleci-v2-sdk

**Comment by devth** [comment]
Why not generate a client from the [CircleCI v2 OpenAPI spec](https://circleci.com/docs/api/v2/)?

**Comment by jithenshriyan** [comment]
@devth I see your point. I've heard though there are some consistency issues with the OpenAPI generator, plus I've included test coverage and reusable types as well. 

I'd like to keep it as a dedicated package for the V2 API, and if you suggest, I can generate the code using OpenAPI and automate releasing whenever the spec changes. What do you think?

**Comment by devth** [comment]
@jithenshriyan sure, up to you. Just seems like a lot of work for something that can be automated 😅

**Comment by jithenshriyan** [comment]
Yeah but I've got most of the major endpoints covered and I'll roll out newer ones if there is a need. It was mostly a project for learning purposes. It still has its advantages with reusable types.

As for the CircleCI tasks within Backstage, I was thinking we could start rolling out changes for the v2 API and I could work on migration with my SDK. Are you also working on tasks for the CircleCI integration as well?

**Comment by freben** [comment]
Putting a proper sdk in place sounds like it might be a plan. To be fully honest, and with all due respect, I suppose it would add some risk to use the linked one - it being an entirely new and not widely used library. As security issues pop up and similar, we of course are eager to have things updated.

Then again - maybe that's an argument for making sure that the circleci functionality stays in a separate module from the core, so that these concerns are localized to that module
